### DISCUSSION TITLE
CheckBox becomes unresponsive to having its checked property set after being manually toggled

### DISCUSSION THREAD

**Discussion by adamski234** [slint-ui/slint#7217] [Q&A]
[Here is the smallest replicable sample of this issue I was able to create](https://slintpad.com/?snippet=import+%7B+CheckBox%2C+Button+%7D+from+%22std-widgets.slint%22%3B%0A%0Aexport+component+AppWindow+inherits+Window+%7B%0A%09in-out+property+%3Cbool%3E+toggled%3B%0A++++VerticalLayout+%7B%0A++++++++cbox+%3A%3D+CheckBox+%7B%0A++++++++++++checked%3A+root.toggled%3B%0A++++++++++++toggled+%3D%3E+%7B%0A++++++++++++++++root.toggled+%3D+self.checked%3B%0A++++++++++++%7D%0A++++++++%7D%0A++++++++Button+%7B%0A++++++++++++text%3A+%22Toggle+CBox%22%3B%0A++++++++++++clicked+%3D%3E+%7B%0A++++++++++++++++root.toggled+%3D+%21root.toggled%3B%0A++++++++++++%7D%0A++++++++%7D%0A++++++++Text+%7B%0A++++++++++++text%3A+%28root.toggled+%3F+%22t%22+%3A+%22f%22%29+%2B+%22+%22+%2B+%28cbox.checked+%3F+%22t%22+%3A+%22f%22%29%3B%0A++++++++%7D%0A++++%7D%0A%7D&style=native). One caveat regarding this snippet: a two-way binding solves this issue, however that does not work for my use case, since I go through a model and they don't support two-way bindings, hence the callback usage.

What happens:
1. Pressing the button toggles the CheckBox, updating both `root.toggled` and `cbox.checked`. The string at the bottom reflects this.
2. Pressing the CheckBox toggles the CheckBox, updating both variables. Due to two-way bindings not being applicable, `root.toggled` is updated via the `toggled` callback from the CheckBox.
3. Pressing the button again does not toggle the CheckBox. `root.toggled` is changed, `cbox.checked` does not change.

What I expect to happen:
At step 3, pressing the button again does toggle the CheckBox, setting `root.toggled` and that being reflected in `cbox.checked` due to the binding. 

Is this behavior expected and I am simply misunderstanding the language, or is this a bug? As far as I can tell, based on the Binding section of the Reactivity page of the documentation, updating `root.toggled` should update `cbox.checked`. The CheckBox page does not mention anything about the bindings being removed on a manual toggle.

**Comment by ogoffart** [comment]
That's because when a value is changed, its binding is broken.
The way to do what you want is with two way bindings:
```slint
            checked <=> root.toggled;
```

**Reply by adamski234** [reply]
Is that mentioned in the docs somewhere and I have missed it?

Also, is the solution something that can be used within repetitions? I'm trying to build a CheckBox tree, composed of roots and their children, where toggling the root sets all the children to the root's toggle state and, if all children are set to the same state, the root gets that state too. Which means that there is some list of checkbox data which get instantiated to CheckBoxes, with some Rust code implementing the above syncing.  
I can't do two-way bindings here, right? I don't see a straightforward way to use functions for that, since Slint doesn't have loops. Though it should be doable with handling it manually from Rust I think, I'd need to try.

For context, [here is the full code that's causing me issues](https://github.com/adamski234/ksr-rs/blob/2e082086ed89cab824f5f225fb7f8520fac9bdc3/ui/checkboxes.slint), in the `CheckBoxTree` component. In a [different file](https://github.com/adamski234/ksr-rs/blob/2e082086ed89cab824f5f225fb7f8520fac9bdc3/ui/app-window.slint) it is instantiated with a two-way binding to a window property, and the property is [manipulated from Rust code](https://github.com/adamski234/ksr-rs/blob/2e082086ed89cab824f5f225fb7f8520fac9bdc3/src/main.rs).
